src/coreclr/vm/assembly.cpp

```diff
@@ -1446,6 +1443,74 @@ INT32 Assembly::ExecuteMainMethod(PTRARRAYREF *stringArgs, BOOL waitForOtherThre
 
             RunManagedStartup();
 
+            Thread::CleanUpForManagedThreadInNative(pThread);
+        }
+    }
+
+    //RunMainPost is supposed to be called on the main thread of an EXE,
+    //after that thread has finished doing useful work.  It contains logic
+    //to decide when the process should get torn down.  So, don't call it from
+    // AppDomain.ExecuteAssembly()
+    if (pMeth) {
+        if (waitForOtherThreads)
+            RunMainPost();
+    }
+    else {
+        StackSString displayName;
+        GetDisplayName(displayName);
+        COMPlusThrowHR(COR_E_MISSINGMETHOD, IDS_EE_FAILED_TO_FIND_MAIN, displayName);
+    }
+
+}
+
+INT32 Assembly::ExecuteMainMethod(PTRARRAYREF *stringArgs, BOOL waitForOtherThreads)
```

Comment:
What are you trying to achieve by this refactoring?

Reply:
What we are trying to determine is the best place to allow swapping the entry assembly and it having a real effect. The goal in https://github.com/dotnet/runtime/issues/99883 is to allow some / all of the TPA assemblies to load, sleep/wake up, and allow the entry assembly to be swapped. 

We've identified two ways to do this. Make the runtime aware, which is what this PR is attempting to do. And the second is another entrypoint in the host that doesn't shut the runtime down after it's done.

Reply:
> * Rely on reflection to load an execute the entry point assembly from the startup hook.
> * Basically the original entry point assembly would never be executed.

The way the runtime appears to work right now is that the original main will be executed regardless of what you do in the hook. Also, aren't there limitations executing code from the hook?

Reply:
> nothing is stopping the hook from calling Environment.Exit

There is a subtle behavior difference vs. returning from `Main`: Environment.Exit does not wait for non-background threads to exit. 

> I think the idea of doing this from the main is better anyway. 

+1

> the entry point replacement as the watch breaks some apps which rely on it.

Yes, a managed API for entry point replacement is generally useful for number of use cases. Something like: https://github.com/jkotas/runtime/commit/a6a0246212fc9548d928e8b7f2459c8ccbd32937 . If you agree, we should start pushing this API through the API review process.

Reply:
> Not sure how this would fit with the host changes to allow for TPA modifications after startup.

TPA additions after startup can be done via the AssemblyLoadContext.Default.

I do not think we should be enabling modifications of TPA after startup. It would a pit of failure when something that wanted to be overridden got loaded too early. Instead, we should figure out how to startup the runtime with TPA subset that is never going to be modified.

Reply:
If we take the Main approach, then the real app main could be invoked by reflection. I wonder what is the need of creating a new API `Assembly.SetEntryPoint`?

Reply:
Without the new API for example this `Assembly.GetEntryAssembly` would return wrong value (it would point to the "host's template `Main`"). I think there are other places that this shows up in some way or form as well. Some apps use this to mostly get to the location of app's files (`Assembly.GetEntryAssembly().Location`) for example.

Reply:
ASP.NET Core has number of places that call Assembly.GetEntryAssembly: https://github.com/search?q=repo%3Adotnet%2Faspnetcore+GetEntryAssembly&type=code

As Vitek said, it is important for Assembly.GetEntryAssembly to return the actual application assembly to avoid breaking these places.

Reply:
Gotcha!